### ISSUE TITLE
Assertion failed: [home_thread() == get_thread_id()]

### ISSUE THREAD

**Issue by larkost** [rethinkdb/rethinkdb#5040]
Locally I am getting a consistent server crash when running `./test/rql_test/test-runner polyglot/changefeeds/sindex.rb1.9` locally. The output is:
```
Version: rethinkdb 2.1.0-0BETA2-403-g0f59e5 (debug) (CLANG 7.0.0 (clang-700.1.76))
error: Error in src/threading.cc at line 27:
error: Assertion failed: [home_thread() == get_thread_id()] 
error: Backtrace:
error: Wed Nov  4 15:34:51 2015
       
       1 [0x10efada90]: 0   rethinkdb                           0x000000010efada90 _Z19rethinkdb_backtracePPvi + 272 at 0x10efada90 ()
       2 [0x10de35370]: 0   rethinkdb                           0x000000010de35370 _ZN11backtrace_tC2Ev + 304 at 0x10de35370 ()
       3 [0x10de371bb]: 0   rethinkdb                           0x000000010de371bb _ZN26lazy_backtrace_formatter_tC2Ev + 43 at 0x10de371bb ()
       4 [0x10de34235]: 0   rethinkdb                           0x000000010de34235 _ZN26lazy_backtrace_formatter_tC1Ev + 21 at 0x10de34235 ()
       5 [0x10de341a0]: 0   rethinkdb                           0x000000010de341a0 _Z16format_backtraceb + 48 at 0x10de341a0 ()
       6 [0x10e89458d]: 0   rethinkdb                           0x000000010e89458d _Z18report_fatal_errorPKciS0_z + 733 at 0x10e89458d ()
       7 [0x10f0f423b]: 0   rethinkdb                           0x000000010f0f423b _ZNK19home_thread_mixin_t13assert_threadEv + 91 at 0x10f0f423b ()
       8 [0x10ea0258c]: 0   rethinkdb                           0x000000010ea0258c _ZN2ql10changefeed14subscription_t4stopESt13exception_ptrNS0_8detach_tE + 44 at 0x10ea0258c ()
       9 [0x10eab6e1c]: 0   rethinkdb                           0x000000010eab6e1c _ZNSt3__128__invoke_void_return_wrapperIvE6__callIJRNS_6__bindIMN2ql10changefeed14subscription_tEFvSt13exception_ptrNS5_8detach_tEEJRNS_12placeholders4__phILi1EEES7_S8_EEEPNS5_11limit_sub_tEEEEvDpOT_ + 700 at 0x10eab6e1c ()
       10 [0x10eab6a5c]: 0   rethinkdb                           0x000000010eab6a5c _ZNSt3__110__function6__funcINS_6__bindIMN2ql10changefeed14subscription_tEFvSt13exception_ptrNS4_8detach_tEEJRNS_12placeholders4__phILi1EEES6_S7_EEENS_9allocatorISE_EEFvPNS4_11limit_sub_tEEEclEOSI_ + 60 at 0x10eab6a5c ()
       11 [0x10ea1b5c2]: 0   rethinkdb                           0x000000010ea1b5c2 _ZNKSt3__18functionIFvPN2ql10changefeed11limit_sub_tEEEclES4_ + 162 at 0x10ea1b5c2 ()
       12 [0x10ea04871]: 0   rethinkdb                           0x000000010ea04871 _ZN2ql10changefeed6feed_t17each_limit_sub_cbERKNSt3__18functionIFvPNS0_11limit_sub_tEEEEi + 1345 at 0x10ea04871 ()
       13 [0x10eaaa264]: 0   rethinkdb                           0x000000010eaaa264 _ZN21pmap_runner_one_arg_tINSt3__16__bindIMN2ql10changefeed6feed_tEFvRKNS0_8functionIFvPNS3_11limit_sub_tEEEEiEJPS4_NS0_17reference_wrapperISA_EERNS0_12placeholders4__phILi1EEEEEExEclEv + 628 at 0x10eaaa264 ()
       14 [0x10eaaa4ac]: 0   rethinkdb                           0x000000010eaaa4ac _ZN26callable_action_instance_tI21pmap_runner_one_arg_tINSt3__16__bindIMN2ql10changefeed6feed_tEFvRKNS1_8functionIFvPNS4_11limit_sub_tEEEEiEJPS5_NS1_17reference_wrapperISB_EERNS1_12placeholders4__phILi1EEEEEExEE10run_actionEv + 28 at 0x10eaaa4ac ()
       15 [0x10de26ebc]: 0   rethinkdb                           0x000000010de26ebc _ZN25callable_action_wrapper_t3runEv + 108 at 0x10de26ebc ()
       16 [0x10de071a5]: 0   rethinkdb                           0x000000010de071a5 _ZN6coro_t3runEv + 917 at 0x10de071a5 ()
error: Exiting.
```

Running from `next` on newton does not reproduce the problem, nor does a freshly built version on gantz. It is possible that this is because of the specific version I happen to have built on my system. Because of a recently fixed build problem I have been using either old, or hacked-up versions.

**Comment by larkost** [comment]
The version of the server is built from 535d37: `2.1.0-0BETA2-398-g535d37 (debug) (CLANG 7.0.0 (clang-700.1.76))` and with that version it reproduces on `gantz` as well.

**Comment by danielmewes** [comment]
We need to investigate this before shipping 2.2.
(mentioning @mlucy as well)

**Comment by danielmewes** [comment]
I can reproduce this on `gantz` even with a fresh build of the current `next` (commit 2ff80bb0aebccc292e0a1ea63a21c8f24a52cbc7 ).

**Comment by danielmewes** [comment]
Alternatively, I'm intermittently seeing this:
```
rethinkdb(25663,0x1139c4000) malloc: *** error for object 0x7fb131f19fe0: incorrect checksum for freed object - object was probably modified after being freed.
*** set a breakpoint in malloc_error_break to debug
```

which might just be a side-effect of data being accessed from the wrong thread.

**Comment by danielmewes** [comment]
... alternatively, the crash about the wrong thread ID might be a side-effect of accessing an already freed object.

**Comment by danielmewes** [comment]
I have what I think is a fix, but it reverts a minor part of the fix for #4972.
I'm going to run the table fuzzer with the configuration mentioned there for a while to make sure that the fix doesn't introduce a regression for the other issue.

**Comment by danielmewes** [comment]
Fix in CR 3315 by @mlucy

**Comment by danielmewes** [comment]
Fixed in next 9800fc1
